### ISSUE TITLE
Registry garbage-collect does not clean up blobs from overwritten tags

### ISSUE THREAD

**Issue by travisgroth** [docker/distribution#2212]
Hi,

We're running registry 2.5.1 with the filesystem storage backend.  No additional flags beyond a secret, certificate path and allowing deletes.

We have a dedicated filesystem for the registry and it has been growing quite large.  We have been pruning (delete + garbage-collection) based on a small program that reports each image/tag's shared and unique blob size total but continue to see growth outpace any usage we can account for.  

Today we simply totaled the size of all the unique blobs; the result was that **the catalog listing only accounted for only 103G out of 686G used** on the physical filesystem.

The methodology for getting the catalog size is:

1) GET /v2/_catalog?n=300 (more than our repo count)
1) for each repository returned, GET /v2/[repository_path]/tags/list
1) for each tag listed, GET /v2/[repository_path]/manifests/[tag]
1) from the manifest returned, HEAD /v2/[repository_path]/blobs/[blob_checksum]
1) store the content-length header

The blobs are stored in a map with their size and then we do a simple sum iterating through the map keys.

Is there some data duplication for v1 compatibility or something else I'm not aware of that might make this methodology flawed?  On an individual image basis, this seems to yield a blob list that approximately equals our predicted size on disk.  The question is, where is the remaining space being used and how can we reclaim it?

Things we've checked for:
- Other applications
- Stale uploads that haven't been purged

Happy to gather more info.

Thanks.

**Comment by travisgroth** [comment]
Update with further info and renaming the issue.

`registry garbage-collect` does **not** clean up old blobs if a tag has been overwritten but has not been deleted based on observations.

Reproduction:

```
export REPO=[myrepo]
dd if=/dev/urandom of=test1 bs=1024000 count=1
dd if=/dev/urandom of=test2 bs=1024000 count=1
echo "FROM scratch
COPY test1 /" > Dockerfile1
echo "FROM scratch
COPY test2 /" > Dockerfile2
docker build -t ${REPO}/testimage -f Dockerfile1 .
docker push ${REPO}/testimage

##
# Check registry blob storage and see new manifest entry and blobs
##
docker build -t ${REPO}/testimage -f Dockerfile2 .
docker push ${REPO}/testimage

##
# Check registry blob storage and see the _manifests/tags/latest/current 'link' points at new hash
# but there are still references to the previous manifest hash under _manifests/tags/latest/ and 
# _manifests/revisions
##

# On Registry:
/usr/bin/registry garbage-collect

# Nothing is GC'ed

```

If I manually delete _manifests/tags/latest/index/sha256/[hash] and _manifests/revisions/sha256/[hash], and then run GC, the old unreachable blob is properly deleted by garbage collection.

It looks like we have almost 600G of data that has been accumulated this way.  

1) Feature request - GC should delete effectively unreachable blobs (at least with a flag to enable this behavior)
2) Question - Is this behavior documented somewhere?
3) Question - Is there a proper way to clean this up without directly inspecting and cleaning up the filesystem for any abandoned manifests still associated with a tag?

Thanks.

**Comment by andrefreitas** [comment]
Hi,
We have the same problem. Our disk usage with the existing images is 40 GB, but old blobs are consuming 374 GB in Disk. 

Since I believe that overwriting tags and using the filesystem storage is a common use case, this is a critical bug that must be solved.

**Comment by dutchakdev** [comment]
Hey, i believed what it works as overwrite too. 
My temporary solutions (looks not fine), bash script + crontab:
`0 0 * * * /opt/clear_old_blobs.sh > /opt/clear-registry.log`
`#!/bin/bash
cd /path/to/blobs/sha256;
let count=0;
for i in $(find -name \*data -mtime +1); do
    rm -rf "$i";
    count++;
done;
find ./ -type d -empty -delete

echo $(date);
echo "Count of removed files: $count"`

i thins it's

**Comment by stevvooe** [comment]
Because images may be referenced by digest or tag, we cannot necessarily collect the untagged images. To have the garbage collection work for this use case, you need to delete the images by digest before running garbage collection.

**Comment by dutchakdev** [comment]
@stevvooe that means if we explicitly specify a tag, for example, some/image:latest then will it be overwritten?

**Comment by jarischaefer** [comment]
If I understand correctly, gargabe collection cannot know which images to delete because one can pull either by image:tag or by digest.

Suppose I regularly push images with the same tag (e.g. development builds). Using pull by image:tag always references the latest image pushed by image:tag, but I could still pull the "overwritten" images by digest. Isn't this similar to dangling images on the docker client?

How can garbage collection not know which images to delete if my use case is simply "keep only the layers needed for image:tag"?

**Comment by stevvooe** [comment]
> Isn't this similar to dangling images on the docker client?

The docker client doesn't really have this problem. I think you mean the daemon. Yes and no. Dangling images (or layers) are objects that are unreferenced by name. Because the collection of dangling images happens on a single node in the daemon, it is much easier to know if something is referenced. With the registry, we have no way to know if someone is referencing an image by digest on a remote system. Simply deleting would break use cases like Docker Content Trust. We allow the user to control this by allowing deletes of manifests, which will then be garbage collected.

> How can garbage collection not know which images to delete if my use case is simply "keep only the layers needed for image:tag"?

Mostly because we err on keeping data over destroying. How do know if you are not referencing it by digest elsewhere? How do you know no one else on your team is using the digest? If this was the default behavior, we'd end up deleting images for a lot of other users. How would that be more fair to them?

My suggestion above is to add an option to garbage collection for users that don't want to hold on to untagged images. This will allow you to run destructive garbage collection for your use case, without deleting data that others may rely on.

**Comment by jarischaefer** [comment]
Thanks for your reply, that's exactly my point. Our team does not reference by digest, thus a "destructive" garbage collection mode would be great to get rid of frequent development builds no one is going to use anyway.

**Comment by gbougeard** [comment]
Hi, is there somewhere a description of registry file tree?
I also have  storage issues and I'd like to clean it.

My use case : 
I have a mutable tag `master` .
I'd like to know what `myimage/_manifests/tag/index/sha256/` hashes are referencing to.
In my mind they are the different "states" of the `master` tag and there should be one hash matching `myimage/_manifests/tag/current/sha256/link` .

And if I'd want to remove the old references I should delete all hashes but the one matching `myimage/_manifests/tag/current/sha256/link` . 
Is that correct?

**Comment by stevvooe** [comment]
@gbougeard The path layout is described in detail [here](https://github.com/docker/distribution/blob/master/registry/storage/paths.go#L21).

> And if I'd want to remove the old references I should delete all hashes but the one matching myimage/_manifests/tag/current/sha256/link .
Is that correct?

Yes, that will work, but make sure that you stop writes to the registry while doing this operation, as the value may change while making a decision about which ones to delete. We have a `readonly` mode option to support this.

> if we have immutable tags, there should be just 1 hash in the index directory (matching the current/link hash) ?

Theoretically, yes, but you'll have to enforce that in your environment, as the registry doesn't do that currently.

**Comment by gbougeard** [comment]
Thanks @stevvooe .

I ran this script I wrote https://gist.github.com/gbougeard/d51bfb6b02a14ae2af7469f60f5b0498 and then ran the GC.
Apparently the GC does not delete all the blobs I listed so the cleaning I did was not as effective as I expected.
For a repo I was waiting to gain 300G but got less than 20G :(

Is my script wrong or is the GC too careful?

**Comment by stevvooe** [comment]
@gbougeard I'm not quite sure, but I don't think you are descending the directory hierarchy at https://gist.github.com/gbougeard/d51bfb6b02a14ae2af7469f60f5b0498#file-remove_outdated_digest-sh-L15. That looks like it will only handle images with a single path component.

**Comment by gbougeard** [comment]
I dont' have namespaces so all images are in `docker/registry/v2/repositories`.
My point is I generated a file with all blobs I think it should be deleted but once I ran the GC if I `ls` those blobs, they are still present.

I based my script on https://gist.github.com/qoomon/7c7f16939630cafafceeb83d254194e4#file-docker-registry-clean-sh-L29 but I did not want to delete the blobs myself like https://gist.github.com/qoomon/7c7f16939630cafafceeb83d254194e4#file-docker-registry-clean-sh-L57 but if the GC does not do it, maybe I have no choice to delete them by myself.

**Comment by stevvooe** [comment]
> if the GC does not do it, maybe I have no choice to delete them by myself.

It's possible that there is a bug. Could you grep for the blob digests you think should be deleted? Maybe there is something that was overlooked?

**Comment by gbougeard** [comment]
for instance my script listed 7308 blobs to be deleted but the GC deleted only 476.

**Comment by stevvooe** [comment]
> for instance my script listed 7308 blobs (for an image) to be deleted but the GC deleted only 476.

Yes, I understand the problem. Check that the blobs that weren't deleted weren't referenced elsewhere. I'd recommend grepping for the blob ids to see if they were referenced somewhere that was unexpected.

**Comment by gbougeard** [comment]
oh, looks like I don't have unique digests in my blobs list to delete :(

**Comment by gbougeard** [comment]
I'm trying this command to find the references of a blob : 
`find docker/registry/v2/blobs/sha256 -type f -exec grep -Iq . {} \; -and -print | xargs grep $BLOB_DIGEST`

The `find` command should list all non-binary files and then `grep` them for the blob digest.
Because I have like 1To of blobs it's very long to process. 

It should be nice to have some storage analysis utilities to be able to find, for example, the tags referencing a blob.

**Comment by gbougeard** [comment]
TIL due to cross referenced blobs, you have to run many times the GC until there are no more blobs eligibles for deletion.

**Comment by stevvooe** [comment]
> EDIT: I'm wondering if https://github.com/docker/distribution/blob/11cb04e994033c5ec98e6630f6e0c6a2b2677344/registry/storage/garbagecollect.go#L87 returns all existing blobs or if it's size limited? I think it is and it would explain why I have to run many times the GC.

This is possible. I would recommend adding a print statement there to see what that method is returning.

**Comment by sasah** [comment]
+1!!!

**Comment by respire** [comment]
Hello, are there any progress on this topic?
Thanks!

**Comment by stevvooe** [comment]
@respire From the latest comments, I think there is enough for a workable feature proposal. We don't have a team to do this work right now, so the community is encouraged to step up. This is a fairly doable task for a contributor, with help.

As far as the next steps are concerned, we'll need the following:

1. Feature proposal: What does the flag do? What does it look like? How does it work? Impact? Nothing long, just an issue here that shows how it works and gives people a change to make commentary.
2. Pull Request: An implementation of the feature based on the proposal. Should be willing to take feedback from contributors and arrive at a solid, safe solution.

Note that this is required to make sure that such a feature avoids having the habit of losing data.

Let me know if you have questions.

**Comment by zetaab** [comment]
@stevvooe I have possible PR solution for this, where I can submit my feature proposal? In github issues or?

**Comment by AndreaGiardini** [comment]
@zetaab Submit a new issue here on github with your feature proposal and eventually attach your PR to it. Let the discussion begin :)

**Comment by RickyCaoCao** [comment]
I ran into a similar problem. Every time a new image is pushed, it took the "latest" tag while keeping its own tag (i.e. "item-version-31"). These tags have different digests so when a newer image is pushed, the digest for the previous "latest" image still exists. Then, the GC does not clean the images because it is still referenced.

I am going to try this script here: https://github.com/mortensteenrasmussen/docker-registry-manifest-cleanup

**Comment by andrefreitas** [comment]
For those that couldn't find a solution to this problem, a Docker Registry in Nexus OSS Repository doesn't suffer from this bug. We have tested overwriting an image multiple times, and cleanup of Nexus OSS works just fine.

https://www.sonatype.com/download-oss-sonatype

**Comment by IngoHeimbach** [comment]
If you use the embedded Docker registry of a GitLab server, you can try [GitLab-Registry-Cleanup](https://github.com/sciapp/gitlab-registry-cleanup).

**Comment by OpakAlex** [comment]
Any news here guys? I have some issue.

**Comment by boianmihailov** [comment]
I got the same problem.

**Comment by vadimk2016** [comment]
For now I use this commands to delete images:
```
docker exec -it registry rm -rf /var/lib/registry/docker/registry/v2/repositories/<NAME>/_manifests
docker exec -it registry  bin/registry garbage-collect  /etc/docker/registry/config.yml
```
But if I delete image from registry, and try to push the same which exists in docker, then registry will say that this image is already exists which is not true, and EOF error will be rised on next pull.

**Comment by zetaab** [comment]
if you use commands like that you need restart the process itself. Instead of rm -rf you should use correct method for deleting manifests and blobs. There is endpoint for those

**Comment by x-yuri** [comment]
@travisgroth @stevvooe This one can probably [be][1] [closed][2]? The feature must be available since `registry-2.7.0` from what I gather.

[1]: https://github.com/docker/distribution/pull/2302#issuecomment-428940952
[2]: https://github.com/docker/distribution/issues/2301#issuecomment-454820398

**Comment by OpakAlex** [comment]
close it, tahnks guys.

**Comment by mibutec** [comment]
As far as I got the issue, deleting images before overriding them would solve the problem.
Would it also be a workaround to iterate over all existing tags for all repositories, delete and recreate them before starting garbage collection?

For me it would be a great help not to have to go to file system for cleanup, so this would be an all-http-solution, is it?

**Comment by mibutec** [comment]
I read the whole thread, but not the latest two comments... Seems 2.7.1 can this without workaround via `-m` option

**Comment by wilsonmar** [comment]
> Because images may be referenced by digest or tag, we cannot necessarily collect the untagged images. To have the garbage collection work for this use case, you need to delete the images by digest before running garbage collection.

Where can we find sample code/commands to achieve that?

**Comment by travisgroth** [comment]
@wilsonmar I think this was closed with https://github.com/docker/distribution/pull/2302.  Going to mark this as such.
